source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+  defined in `createEntities`_ but may also be a
+  `saved operation result <operation_saveResultAsEntity_>`_. Entities are
+  referenced throughout the test file (e.g. `Entity Test Operations`_).
+
+Internal MongoClient
+  A MongoClient created specifically for use with internal test operations, such
+  as inserting collection data before a test, configuring fail points during a
+  test, or asserting collection data after a test.
+
+
+Schema Version
+--------------
+
+This specification and the `Test Format`_ follow
+`semantic versioning <https://semver.org/>`__. Backwards breaking changes (e.g.
+removing a field, introducing a required field) will warrant a new major
+version. Backwards compatible changes (e.g. introducing an optional field) will
+warrant a new minor version. Small fixes and internal changes (e.g. grammar,
+adding clarifying text to the spec) will warrant a new patch version; however,
+patch versions SHOULD NOT alter the structure of the test format and thus SHOULD
+NOT be relevant to test files.
+
+Each test file defines a `schemaVersion`_, which test runners will use to
+determine compatibility (i.e. whether and how the test file will be
+interpreted). Test files are considered compatible with a test runner if their
+`schemaVersion`_ is less than or equal to a supported version in the test
+runner, given the same major version component. For example:
+
+- A test runner supporting version 1.5.1 could execute test files with versions
+  1.0 and 1.5 but *not* 1.6 and 2.0.
+- A test runner supporting version 2.1 it could execute test files with versions
+  2.0 and 2.1 but *not* 1.0 and 1.5.
+- A test runner supporting *both* versions 1.5.1 and 2.0 could execute test
+  files with versions 1.4, 1.5, and 2.0, but *not* 1.6, 2.1, or 3.0.
+- A test runner supporting version 2.0.1 could execute test files with versions
+  2.0 and 2.0.1 but *not* 2.0.2 or 2.1. This example is provided for
+  completeness, but test files SHOULD NOT need to refer to patch versions (as
+  previously mentioned).
+
+Test runners MUST NOT process incompatible files but MAY determine how to handle
+such files (e.g. skip and log a notice, fail and raise an error). Test runners
+MAY support multiple schema versions (as demonstrated in the example above).
+
+Each major version of this specification will have a corresponding JSON schema
+(e.g. `schema-1.json <schema-1.json>`__), which may be used to programmatically
+validate YAML and JSON files using a tool such as `Ajv <https://ajv.js.org/>`__.
+
+The latest JSON schema MUST remain consistent with the `Test Format`_ section.
+If and when a new major version is introduced, the `Breaking Changes`_ section
+must be updated and JSON schema(s) for any previous major version(s) MUST remain
+available so that older test files can still be validated. New tests files
+SHOULD always be written using the latest version of this specification.
+
+
+Entity Map
```

Comment:
Currently `createEntities` is an array of keyed objects.

It's my suggestion that this is just an array of objects.

Currently:

```js
{
  createEntities: [
    {
      client: {
        id: 'client-id',
      }
    }
  ]
}
```

Suggestion:

```js
{
  createEntities: [
    {
      name: 'client',
      id: 'client-id',
    }
  ]
}
```

This would mimic or gain consistency with the operation and name.

There is no real guard against single-key objects, meaning a way of stopping.
This doesn't read very well. It would need an extra guard to check.

Validating single-keyed object:

This is what the doc says we must not do, but there's no real reason why not.

```js
{
  createEntities: [
    {
      client: {
        id: string;
        observeEvents: string[];
      },
      database: {
        id: string;
        observeEvents: string[];
      }
    },
    {
      client: {
        id: string;
        observeEvents: string[];
      },
    }
  ],
}
```

Reply:
As I mentioned in a previous comment validating using `{ "minProperties": 1, "maxProperties": 1 }` sounds great.